src/s2geography/geography.h

```diff
@@ -170,31 +235,128 @@ class GeographyCollection : public Geography {
 // valid for the scope of those objects.
 class ShapeIndexGeography : public Geography {
  public:
-  ShapeIndexGeography(
-      MutableS2ShapeIndex::Options options = MutableS2ShapeIndex::Options())
-      : shape_index_(options) {}
+  ShapeIndexGeography();
+  ShapeIndexGeography(int max_edges_per_cell);
 
-  explicit ShapeIndexGeography(const Geography& geog) { Add(geog); }
+  explicit ShapeIndexGeography(const Geography& geog);
 
   // Add a Geography to the index, returning the last shape_id
   // that was added to the index or -1 if no shapes were added
   // to the index.
-  int Add(const Geography& geog) {
-    int id = -1;
-    for (int i = 0; i < geog.num_shapes(); i++) {
-      id = shape_index_.Add(geog.Shape(i));
-    }
-    return id;
-  }
+  int Add(const Geography& geog);
+
+  int num_shapes() const;
+  std::unique_ptr<S2Shape> Shape(int id) const;
+  std::unique_ptr<S2Region> Region() const;
+
+  const S2ShapeIndex& ShapeIndex() const { return *shape_index_; }
+
+  void Encode(Encoder* encoder, const EncodeOptions& options) const;
+
+ private:
+  std::unique_ptr<MutableS2ShapeIndex> shape_index_;
+};
+
+// A Geography with a EncodedS2ShapeIndex as the underlying data.
+// This is to facilitate decoding, whereas a MutableS2ShapeIndex is
+// used to construct the S2ShapeIndex required for many S2 operations.
+class EncodedShapeIndexGeography : public Geography {
+ public:
+  EncodedShapeIndexGeography();
 
   int num_shapes() const;
   std::unique_ptr<S2Shape> Shape(int id) const;
   std::unique_ptr<S2Region> Region() const;
 
-  const MutableS2ShapeIndex& ShapeIndex() const { return shape_index_; }
+  const S2ShapeIndex& ShapeIndex() const { return *shape_index_; }
+
+  void Encode(Encoder* encoder, const EncodeOptions& options) const;
+
+  void Decode(Decoder* decoder, const EncodeOptions& options);
 
  private:
-  MutableS2ShapeIndex shape_index_;
+  std::unique_ptr<S2ShapeIndex> shape_index_;
+  std::unique_ptr<S2ShapeIndex::ShapeFactory> shape_factory_;
+};
+
+// Options for serializing geographies using Geography::EncodeTagged()
+class EncodeOptions {
+ public:
+  // Create options with default values, which optimize for the
+  // scenario where a geography is getting serialized and hopefully
+  // not deserialized in full.
+  EncodeOptions() = default;
+
+  // Control whether to optimize for speed (by writing vertices as
+  // doubles) or space (by writing cell identifiers for vertices that
+  // are snapped to a cell center). For vertices that are snapped to a
+  // cell center at a lower zoom level, the encoder can encode each
+  // vertex with 4 or fewer bytes.
+  void set_coding_hint(s2coding::CodingHint hint) { hint_ = hint; }
+  s2coding::CodingHint coding_hint() const { return hint_; }
+
+  // Control whether to spend extra effort converting shapes that
+  // aren't able to be lazily decoded (e.g., S2Polyline::Shape and
+  // S2Polygon::Shape to S2LaxPolylineShape and S2LaxPolygonShape,
+  // respectively).
```

Comment:
So by default several of the geography types can be decoded lazily?  
And the "lazy decoding" means that you create eg a S2Polyline object where it somehow still points to the serialized blob and only decodes when needed?